common/achievement/evaluate.ts

```diff
@@ -0,0 +1,114 @@
+import { Achievement_event } from '../../graphql/generated';
+import { AchievementContextType, BucketConfig, BucketEvents, ConditionDataAggregations, EvaluationResult } from './types';
+import { prisma } from '../prisma';
+import { aggregators } from './aggregator';
+import swan from '@onlabsorg/swan-js';
+import { bucketCreatorDefs } from './bucket';
+import { getLogger } from '../logger/logger';
+import { getBucketContext } from './util';
+const logger = getLogger('Achievement');
+
+export async function evaluateAchievement(
+    condition: string,
+    dataAggregation: ConditionDataAggregations,
+    metrics: string[],
+    recordValue: number
+): Promise<EvaluationResult> {
+    const achievementEvents = await prisma.achievement_event.findMany({ where: { metric: { in: metrics } }, orderBy: { createdAt: 'desc' } });
+
+    const eventsByMetric: Record<string, Achievement_event[]> = {};
+    for (const event of achievementEvents) {
+        if (!eventsByMetric[event.metric]) {
+            eventsByMetric[event.metric] = [];
+        }
+        eventsByMetric[event.metric].push(event);
+    }
+
+    const resultObject: Record<string, number> = {};
+    resultObject['recordValue'] = recordValue;
+
+    for (const key in dataAggregation) {
+        if (!dataAggregation[key]) {
+            continue;
+        }
+        const dataAggregationObject = dataAggregation[key];
+        const metricName = dataAggregationObject.metric;
+
+        const bucketCreator = dataAggregationObject.createBuckets || 'default';
+        const bucketAggregator = dataAggregationObject.bucketAggregator || 'count';
+
+        const aggregator = dataAggregationObject.aggregator;
+
+        const eventsForMetric = eventsByMetric[metricName];
+        if (!eventsForMetric) {
+            continue;
+        }
```

Comment:
I think I would remove this. It doesn't hurt to evaluate an empty array, and maybe we have reasons to do it.